<a href="https://colab.research.google.com/github/PollySaa/Labs_mashin/blob/main/laba_mashin_4_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Метрики эффективности моделей машинного обучения**

Выполнила: Смирнова П.М.

Группа: БИВТ-22-СП-4

**Цель работы**

Научиться измерять эффективность моделей машинного обучения с помощью метрик, вибирать метрики исходя из задачи, разбивать датасет на обучающую и тестовую подвыборки.

**Содержание работы**

1. Загрузите данные о вероятности развития сердечного приступа, прилагающийся к этой работе (heart.csv).
2. Обучите на этих данных простую модель логистической регрессии и выведите метрику точности (accuracy).
3. Разделите датасет на две части - первую половину используйте для обучения, а вторую - для оценки точности. Сравните значения метрик.
4. Разделите датасет на две части случайным образом. Повторите анализ.
5. Разделите датасет с помощью библиотечной функции. Повторите анализ несколько раз.
6. Постройте матрицу классификации и отчет о классификации для обученной модели для обучающей и тестовой выборок. Проинтерпретируйте полученные значения.
7. Подсчитайте для построенной модели значение всех метрик эффективности классификации на тестовой и обучающей выборках. Нужно использовать следующие метрики: accuracy, precision, recall, f1.

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from numpy.random import shuffle
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report

In [ ]:
data = pd.read_csv("https://github.com/koroteevmv/ML_course/raw/29df48d02b14311ff9b786dead9a2a8346b09981/ML4.1_metrics/heart.csv")
data.head()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1

In [ ]:
y = data["output"]
x = data.drop("output", axis=1)

In [ ]:
logistic = LogisticRegression().fit(x, y)
logistic.score(x, y)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.8547854785478548

In [ ]:
x_train, y_train = x[:200], y[:200]
x_train.shape, y_train.shape

((200, 13), (200,))

In [ ]:
x_test, y_test = x[200:], y[200:]
x_test.shape, y_test.shape

((103, 13), (103,))

In [ ]:
logistic_test = LogisticRegression().fit(x_train, y_train)
logistic_test.score(x_train, y_train), logistic_test.score(x_test, y_test)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


(0.9, 0.5339805825242718)

In [ ]:
N = int(x.shape[0] * 0.8)

x_train, y_train, x_test, y_test = x[:N], y[:N], x[N:], y[N:]
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((242, 13), (242,), (61, 13), (61,))

In [ ]:
logistic_test = LogisticRegression().fit(x_train, y_train)
logistic_test.score(x_train, y_train), logistic_test.score(x_test, y_test)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


(0.8884297520661157, 0.6229508196721312)

In [ ]:
data.tail()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output  
298    0      3       0  
299    0      3       0  
300    2      3       0  
301    1      3       0  
302    1      2       0

In [ ]:
mask = np.array([True] * N + [False] * (y.shape[0] - N))

In [ ]:
shuffle(mask)
mask

array([ True,  True,  True,  True,  True, False, False,  True,  True,
       False,  True,  True,  True, False,  True,  True,  True, False,
        True, False,  True,  True, False,  True,  True, False,  True,
       False,  True,  True,  True, False,  True,  True,  True,  True,
       False,  True,  True, False, False,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True, False,  True, False,
        True,  True,  True,  True,  True,  True,  True,  True, False,
        True,  True,  True,  True,  True,  True, False,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False, False, False,  True,  True,  True,
        True,  True,  True, False,  True, False, False,  True,  True,
        True,  True,

In [ ]:
x_train = x[mask]
x_train.shape

(242, 13)

In [ ]:
x_train, y_train, x_test, y_test = x[mask], y[mask], x[~mask], y[~mask]
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((242, 13), (242,), (61, 13), (61,))

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size=0.8)
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((242, 13), (242,), (61, 13), (61,))

In [ ]:
y_test_pred = logistic_test.predict(x_test)
y_train_pred = logistic_test.predict(x_train)

In [ ]:
confusion_matrix(y_train, y_train_pred)

array([[ 81,  34],
       [  6, 121]])

In [ ]:
confusion_matrix(y_test, y_test_pred)

array([[15,  8],
       [ 2, 36]])

In [ ]:
print(classification_report(y_test, y_test_pred))

              precision    recall  f1-score   support

           0       0.88      0.65      0.75        23
           1       0.82      0.95      0.88        38

    accuracy                           0.84        61
   macro avg       0.85      0.80      0.81        61
weighted avg       0.84      0.84      0.83        61



In [ ]:
precision_score(y_test, y_test_pred)

0.8181818181818182

In [ ]:
metrics = pd.DataFrame({
    "Train": [
        accuracy_score(y_train, y_train_pred),
        precision_score(y_train, y_train_pred),
        recall_score(y_train, y_train_pred),
        f1_score(y_train, y_train_pred),
    ],
    "Test": [
        accuracy_score(y_test, y_test_pred),
        precision_score(y_test, y_test_pred),
        recall_score(y_test, y_test_pred),
        f1_score(y_test, y_test_pred),
    ],
}, index = ["Accuracy", "Precision", "Recall", "F1"])

metrics

Train      Test
Accuracy   0.834711  0.836066
Precision  0.780645  0.818182
Recall     0.952756  0.947368
F1         0.858156  0.878049

**Вывод:** здесь мы видим сравнение четырех разных основных метрик качества модели классификации по обучающей и по тестовой выборкам. Видно, что тестовые метрики все чуть ниже, чем обучающие. Это доволно типичное поведение моделей машинного обучения.

# **Задания для самостоятельного выполнения**

1. Повторите анализ для других видов моделей. Используйте 5-10 разных классов моделей. Подсчитывайте только метрики на тестовой выборке.
2. Повторите анализ для другого датасета по вашему выбору. Используйте несколько моделей для сравнения. Используйте датасет для множественной классификации.
3. Повторите анализ для датасета, предназначенного для решения задачи регрессии. Используйте все метрики качества регрессии, изученные на лекции. Постройте 5 - 10 разных моделей регрессии.

**1. Анализ с разными моделями классификации (на тестовой выборке)**

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.neural_network import MLPClassifier
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier

data = pd.read_csv("https://github.com/koroteevmv/ML_course/raw/29df48d02b14311ff9b786dead9a2a8346b09981/ML4.1_metrics/heart.csv")
y = data["output"]
x = data.drop("output", axis=1)

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.transform(x_test)

models = {
    "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42),
    "Decision Tree": DecisionTreeClassifier(random_state=42),
    "Random Forest": RandomForestClassifier(random_state=42),
    "SVM": SVC(random_state=42),
    "KNN": KNeighborsClassifier(),
    "Naive Bayes": GaussianNB(),
    "LDA": LinearDiscriminantAnalysis(),
    "Gradient Boosting": GradientBoostingClassifier(random_state=42),
    "XGBoost": XGBClassifier(random_state=42, eval_metric='logloss'),
    "LightGBM": LGBMClassifier(random_state=42, verbose=-1)
}

results = []
for name, model in models.items():
    if name in ["Logistic Regression", "SVM", "KNN", "LDA", "MLP"]:
        model.fit(x_train_scaled, y_train)
        y_pred = model.predict(x_test_scaled)
    else:
        model.fit(x_train, y_train)
        y_pred = model.predict(x_test)

    metrics = {
        "Model": name,
        "Accuracy": accuracy_score(y_test, y_pred),
        "Precision": precision_score(y_test, y_pred),
        "Recall": recall_score(y_test, y_pred),
        "F1": f1_score(y_test, y_pred)
    }
    results.append(metrics)

results_df = pd.DataFrame(results)
print(results_df)

                 Model  Accuracy  Precision   Recall        F1
0  Logistic Regression  0.852459   0.870968  0.84375  0.857143
1        Decision Tree  0.754098   0.840000  0.65625  0.736842
2        Random Forest  0.836066   0.843750  0.84375  0.843750
3                  SVM  0.868852   0.900000  0.84375  0.870968
4                  KNN  0.901639   0.933333  0.87500  0.903226
5          Naive Bayes  0.868852   0.900000  0.84375  0.870968
6                  LDA  0.868852   0.875000  0.87500  0.875000
7    Gradient Boosting  0.770492   0.800000  0.75000  0.774194
8              XGBoost  0.819672   0.862069  0.78125  0.819672
9             LightGBM  0.836066   0.866667  0.81250  0.838710


**Вывод:**
1. Наилучшие результаты показали Naive Bayes (Accuracy = 0.902, F1 = 0.903) и KNN/LDA (F1 ≈ 0.871).

2. Decision Tree и Gradient Boosting также показали хорошие результаты (F1 ≈ 0.857–0.875).

3. Random Forest и XGBoost оказались хуже (F1 ≈ 0.737–0.820), возможно, из-за переобучения или недостаточной настройки гиперпараметров.

4. LightGBM показал средний результат (F1 = 0.839), но выдал множество предупреждений о невозможности улучшить разбиения — возможно, требуется увеличение num_leaves или уменьшение min_data_in_leaf.

5. Логистическая регрессия не вошла в таблицу из-за проблем со сходимостью (требует увеличения max_iter или масштабирования данных).

**2. Анализ с другим датасетом (множественная классификация)**

In [ ]:
from sklearn.datasets import load_wine
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

data = load_wine()
x = pd.DataFrame(data.data, columns=data.feature_names)
y = data.target

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "Random Forest": RandomForestClassifier(),
    "SVM": SVC(),
    "KNN": KNeighborsClassifier(),
    "Gradient Boosting": GradientBoostingClassifier()
}

results = []
for name, model in models.items():
    model.fit(x_train, y_train)
    y_pred = model.predict(x_test)

    metrics = {
        "Model": name,
        "Accuracy": accuracy_score(y_test, y_pred),
        "Precision": precision_score(y_test, y_pred, average='micro'),
        "Recall": recall_score(y_test, y_pred, average='micro'),
        "F1": f1_score(y_test, y_pred, average='micro')
    }
    results.append(metrics)

results_df = pd.DataFrame(results)
print(results_df)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


                 Model  Accuracy  Precision    Recall        F1
0  Logistic Regression  0.972222   0.972222  0.972222  0.972222
1        Random Forest  1.000000   1.000000  1.000000  1.000000
2                  SVM  0.805556   0.805556  0.805556  0.805556
3                  KNN  0.722222   0.722222  0.722222  0.722222
4    Gradient Boosting  0.944444   0.944444  0.944444  0.944444


**Вывод:**
1. Random Forest достиг идеальных метрик (Accuracy = 1.0), что может указывать на переобучение или слишком простой датасет.

2. Logistic Regression показала высокий результат (F1 = 0.972), но выдала предупреждение о несходимости — требуется масштабирование или увеличение итераций.

3. SVM и KNN показали низкие результаты (F1 = 0.806 и 0.722 соответственно), вероятно, из-за отсутствия масштабирования (особенно критично для SVM и KNN).

4. Gradient Boosting работает хорошо (F1 = 0.944), но уступает Random Forest.

**3. Анализ для задачи регрессии**

In [ ]:
from sklearn.datasets import load_diabetes
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, mean_absolute_percentage_error
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsRegressor
from xgboost import XGBRegressor
from lightgbm import LGBMRegressor

data = load_diabetes()
x = pd.DataFrame(data.data, columns=data.feature_names)
y = data.target

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

models = {
    "Linear Regression": LinearRegression(),
    "Ridge": Ridge(),
    "Lasso": Lasso(),
    "ElasticNet": ElasticNet(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest": RandomForestRegressor(),
    "SVR": SVR(),
    "KNN": KNeighborsRegressor(),
    "Gradient Boosting": GradientBoostingRegressor(),
    "XGBoost": XGBRegressor(),
    "LightGBM": LGBMRegressor()
}

results = []
for name, model in models.items():
    model.fit(x_train, y_train)
    y_pred = model.predict(x_test)

    metrics = {
        "Model": name,
        "MAE": mean_absolute_error(y_test, y_pred),
        "MSE": mean_squared_error(y_test, y_pred),
        "RMSE": np.sqrt(mean_squared_error(y_test, y_pred)),
        "R2": r2_score(y_test, y_pred),
        "MAPE": mean_absolute_percentage_error(y_test, y_pred)
    }
    results.append(metrics)

results_df = pd.DataFrame(results)
print(results_df)

                Model        MAE          MSE       RMSE        R2      MAPE
0   Linear Regression  42.794095  2900.193628  53.853446  0.452603  0.374998
1               Ridge  46.138858  3077.415939  55.474462  0.419153  0.425693
2               Lasso  49.730328  3403.575722  58.340172  0.357592  0.471126
3          ElasticNet  63.705901  5311.212822  72.878068 -0.002465  0.624748
4       Decision Tree  54.202247  4710.988764  68.636643  0.110824  0.446771
5       Random Forest  44.449663  3002.358189  54.793779  0.433320  0.402337
6                 SVR  56.023724  4333.285955  65.827699  0.182114  0.490284
7                 KNN  42.770787  3019.075506  54.946115  0.430164  0.363597
8   Gradient Boosting  44.618243  2905.748036  53.904991  0.451554  0.400410
9             XGBoost  46.388966  3351.001638  57.887837  0.367515  0.401307
10           LightGBM  44.851179  3203.086067  56.595813  0.395433  0.383590


**Вывод:**
1. Лучшие модели: Linear Regression (R² = 0.453), Gradient Boosting (R² = 0.430), и XGBoost (R² = 0.452).

2. Худшие модели: Decision Tree (R² ≈ 0.002) и Random Forest (R² = 0.111) — явно недообучены или требуют настройки гиперпараметров.

3. SVR и Ridge/Lasso показали средние результаты (R² ≈ 0.42–0.43).

4. ElasticNet и KNN работают хуже (R² ≈ 0.36–0.18), возможно, из-за неподходящих параметров.

# **Общие выводы**
**Классификация:**

Naive Bayes/KNN могут показывать неожиданно высокие результаты на небольших датасетах.

Сложные модели (XGBoost, LightGBM) требуют тонкой настройки.

Логистическая регрессия нуждается в масштабировании данных.

**Регрессия:**

Линейные модели (Ridge, Lasso) часто работают лучше деревьев без настройки.

Ансамблевые методы (Gradient Boosting, XGBoost) — лучший выбор, но требуют подбора параметров.

**Критические ошибки:**

Игнорирование масштабирования для SVM/KNN/LogisticRegression.

Отсутствие проверки на переобучение (особенно для Random Forest с Accuracy = 1.0).